### PR TITLE
Tuples: Add some tests on interface implementation

### PR DESCRIPTION
**Pull request by jcouv**
I suspect the IDE crash I encountered with partial interface code was due to one of the samples installed in my hive. But I didn't get a stack trace and now issue #10569 doesn't repro.
Adding a few tests for interfaces involving tuples.

@VSadov @dotnet/roslyn-compiler for review.

### CODE REVIEW

**Comment by jcouv** [comment]
@VSadov Could you review?

**Comment by VSadov** [comment]
I'd add couple generic cases like:

```C#
interface ISomething<T, U> where T: U
{
       (T, U, V) M<V>((T, (U, V)) arg) where V: T;
}
```

**Comment by VSadov** [comment]
I tried some cases in IDE as well and I do not see a crash. Perhaps it was something else - sample analyzer, or the other thing that has been fixed...

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 1]
```diff
@@ -4139,5 +4139,84 @@ static void Main()
...
+interface I
+{
+    (int Alice, string Bob) M((int x, string y) value);
+    (int Alice, string Bob) P1 { get; }
+}
+class C : I
+{
+    static void Main()
+    {
+        var c = new C();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: Perhaps this should be typed as an interface so that calls go through it?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: I see this done below.

**Comment by AlekseyTs** [comment]
We probably should also test scenarios when class implements a generic interface with tuple type as a type argument.
Another scenario I am curious about is when base class and derived class implement the same generic interface, but with tuple types differ only by name as type arguments. What AllInterfaces API is going to return in this case?

**Comment by jcouv** [comment]
@VSadov I added the test you suggested.
Note that we already had a bunch of cases on interfaces: VT arrays in and out, tuples with names/without, long tuples in and out, tuple implementation of VT interface, tuple implementation of VT interface.

**Comment by AlekseyTs** [comment]
Current tests LGTM

**Comment by jcouv** [comment]
@AlekseyTs Good ideas. I'll add those tests. Thanks

**Comment by VSadov** [comment]
Considering that this is just an attempt to find a repro for a crash in a specific scenario, this 
LGTM.

**Comment by jcouv** [comment]
@AlekseyTs I added some tests for the scenarios you described. But I didn't under what you meant by "AllInterfaces API". 
Let me know if you think we should explicitly check some symbols.

**Comment by AlekseyTs** [comment]
Yes, this is an API on a TypeSymbol, it would be interesting to see what does it return for the derived type.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 2]
```diff
@@ -4296,5 +4296,100 @@ static void Main()
...
+
+class C : I<(int b, int a), (int a, int b)>
+{
+    public virtual ((int, int) x, (int, int) y) M(((int, int), (int, int)) x)
+    {
+        throw new System.Exception();
+    }
+}
+
+class D : C
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: The scenario I was talking about is
```
class Base : I<(int a, int b)>
...
class Derived : Base, I<(int c, int d)>
```

**Comment by jcouv** [comment]
@AlekseyTs Added the AllInterfaces check with the scenario you clarified. Two interfaces are returned.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 3]
```diff
@@ -4389,7 +4389,19 @@ public override ((int b, int a), (int b, int a)) M(((int a, int b), (int b, int
...
-            var comp = CompileAndVerify(source, expectedOutput: @"{1, 2} {3, 4}", parseOptions: TestOptions.Regular.WithTuplesFeature());
+            Action<ModuleSymbol> validator = module =>
+            {
+                var sourceModule = (SourceModuleSymbol)module;
+                var compilation = sourceModule.DeclaringCompilation;
+                TypeSymbol y = (TypeSymbol)compilation.GlobalNamespace.GetMember("D");
+
+                Assert.Equal(2, y.AllInterfaces.Length);
+                Assert.Equal("I<(System.Int32 b, System.Int32 a), (System.Int32 a, System.Int32 b)>", y.AllInterfaces[0].ToTestDisplayString());
+                Assert.Equal("I<(System.Int32 a, System.Int32 b), (System.Int32 c, System.Int32 d)>", y.AllInterfaces[1].ToTestDisplayString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: Please open an issue to confirm that the current behavior is what we expect.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs (hunk 3 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: Opened issue #10923

**Pull request merged**